pkg/volume/csi/csi_mounter.go

```diff
@@ -375,6 +375,15 @@ func (c *csiMountMgr) TearDown() error {
 func (c *csiMountMgr) TearDownAt(dir string) error {
 	klog.V(4).Infof(log("Unmounter.TearDownAt(%s)", dir))
 
+	// Deprecation: Removal of target_path provided in the NodePublish RPC call
+	// (in this case location `dir`) MUST be done by the CSI plugin according
+	// to the spec. This will no longer be done directly as part of TearDown
+	// by the kubelet in the future. Kubelet will only be responsible for
+	// removal of json data files it creates and parent directories.
+	if err := removeMountDir(c.plugin, dir); err != nil {
+		return errors.New(log("Unmounter.TearDownAt failed to clean mount dir [%s]: %v", dir, err))
+	}
+
 	volID := c.volumeID
```

Comment:
This does not seem correct, so if `NodeUnpublish` fails then subsequent calls to `NodeUnpublish`  won't even start because directory and json files has already been removed. 

I do not think fixing it here is the right solution but we need to think it from perspective of how cleanup should work if drivers are not *yet* available or volume is not yet mounted.  IMO - the fix has to be made in reconstruction code, not here.

Reply:
Actually, csi mount directory will be removed anyway by housekeeping handler in [removeOrphanedPodVolumeDirs](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_volumes.go#L119), this function will even remove entire pod volume dir, which is `/var/lib/kubelet/pods/<POD_UID>/volumes`, the reason why the directory was not deleted sucessfully and left over is that there is a file called `vol_data.json`, and this function can only delete directory but not file. So here I put `removeMountDir` before calling out csi driver to make sure `vol_data.json` is cleaned up no matter `NodeUnpublish` sucess or not, in this way, the orphan pod volume will be cleaned up by housekeeping handler eventually.

Reply:
I agree with @gnufied , this doesn't seem like the right place to fix it.
What happens when the directory is still mounted? Removal of the directory would fail then but with this change it means `TearDownAt` would return an error before calling `NodeUnpublishVolume` to unmount the volume, and it would never get unmounted. So I think this breaks the main use case (unmount volume and then remove the mountpoint).

Reply:
Actually, when the directory is still mounted, `removeMountDir` won't return error. It still should be unmounted by csi driver first. Every cleaning up action in kubelet should have considered the mount state in the first place.

Considering that csi driver is responsible for unmounting and cleaning the `mount` directory, maybe we could only remove `vol_data.json` here, which should be done by kubelet, rather than remove the mount dir before calling out `NodeUnpublishVolume`.

Reply:
@gnufied @dobsonj I agree that we should not break the original logic of `TearDownAt`. And I update the PR. This time I put the deletion of the volume data file in kubelet housekeeping process. Could you PTAL?